components/odh-notebook-controller/controllers/notebook_webhook.go

```diff
@@ -438,3 +450,106 @@ func InjectCertConfig(notebook *nbv1.Notebook, configMapName string) error {
+                return nil
+            } else {
+                // Split the imageSelection to imagestream and tag
+                parts := strings.Split(imageSelection, ":")
+                if len(parts) != 2 {
+                    log.Error(nil, "Invalid image selection format")
+                    return fmt.Errorf("invalid image selection format")
+                }
+
+                imagestreamName := parts[0]
+                tag := parts[1]
+
+                // Specify the namespaces to search in
+                namespaces := []string{"opendatahub", "redhat-ods-applications"}
+
+                imagestreamFound := false
+
+                for _, namespace := range namespaces {
+                    // List imagestreams in the specified namespace
+                    imagestreams, err := dynamicClient.Resource(ims).Namespace(namespace).List(ctx, metav1.ListOptions{})
+                    if err != nil {
+                        log.Error(err, "Cannot list imagestreams", "namespace", namespace)
+                        continue
+                    }
+
+                    // Iterate through the imagestreams to find matches
+                    for _, item := range imagestreams.Items {
+                        metadata := item.Object["metadata"].(map[string]interface{})
+                        name := metadata["name"].(string)
+
+                        if name == imagestreamName {
+                            status := item.Object["status"].(map[string]interface{})
+
+                            log.Info("No Internal registry found, pick up imageHash from status.tag.dockerImageReference")
+
+                            tags := status["tags"].([]interface{})
+                            for _, t := range tags {
+                                tagMap := t.(map[string]interface{})
+                                tagName := tagMap["tag"].(string)
+                                if tagName == tag {
+                                    items := tagMap["items"].([]interface{})
+                                    if len(items) > 0 {
+                                        // Sort items by creationTimestamp to get the most recent one
+                                        sort.Slice(items, func(i, j int) bool {
+                                            iTime := items[i].(map[string]interface{})["created"].(string)
+                                            jTime := items[j].(map[string]interface{})["created"].(string)
+                                            return iTime > jTime // Lexicographical comparison of RFC3339 timestamps
+                                        })
+                                        imageHash := items[0].(map[string]interface{})["dockerImageReference"].(string)
+                                        notebook.Spec.Template.Spec.Containers[0].Image = imageHash
+                                        // Update the JUPYTER_IMAGE environment variable
```

Comment:
Hi all, definitely, putting an sha256 style name in, especially for an image, not an imagestream, is not ok.
What I can say for sure is, that, during my work on PR-800, I found out that JUPYTER_IMAGE env var is used during commercial RHOIA / upstream CI testing to determine imagestream name and tag. 
https://github.com/red-hat-data-services/ods-ci/blob/2c62e207183f1ec560ca44cefdcfa0bfb83c0c4d/ods_ci/tests/Resources/Page/ODH/JupyterHub/JupyterHubSpawner.robot#L305

Reply:
Also, as I mentioned before, ODH notebook controller is the wrong place to handle this sort of stuff.
Much better in my view to start declaratively with odh dashboard and the image change trigger annotation.